llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -745,6 +748,33 @@ static bool areTypesABICompatible(ArrayRef<Type *> Types, const Function &F,
   });
 }
 
+// Try to prove that all Calls to F do not modify the memory pointed to by Arg.
+// This can provide us with more opportunities to perform Argument Promotion in
+// cases where simply looking at a Function's instructions is insufficient to
+// prove that the pointer argument is not invalidated before all loads from it.
+static bool callDoesNotModifyArg(Function *F, unsigned ArgNo,
+                                 FunctionAnalysisManager &FAM) {
+  // Find all Users of F that are Calls, and see if they may modify Arg.
+  for (User *U : F->users()) {
+    auto *Call = dyn_cast<CallInst>(U);
+    if (!Call)
+      continue;
+
+    Value *ArgOp = Call->getArgOperand(ArgNo);
+    assert(ArgOp->getType()->isPointerTy() && "Argument must be Pointer Type!");
+
+    MemoryLocation Loc = MemoryLocation::getForArgument(Call, ArgNo, nullptr);
```

Comment:
Can we construct a more precise memory location based on the actual loads in the callee?  This is basically equivalent to `MemoryLocation::getBeforeOrAfter(ArgOp)`

Reply:
If I understand correctly, I don't think that we can do this - because BasicAA does not support inter-procedural analysis, we cannot pass a Loc derived from a load in the callee to the AAR query based on the caller (where the Call to F is).
https://github.com/llvm/llvm-project/blob/6c143a86cddbc6d0431dd643bfc7d4f017042512/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L925-L929

Reply:
I mean, use the pointer from the caller, but compute the size of the memory region using the loads in the callee.

Reply:
Ah I see, apologies I was misunderstanding you before. I have added this in [Address review 3](https://github.com/llvm/llvm-project/pull/106216/commits/6b9b01890f86389212cbe762e43bf896215dd9dd)

If I understand correctly we still run into an issue because we are doing an alias query on a CallInst, and because BasicAA is inherently non-interprocedural it disregards the Size of the Loc and uses getBeforeOrAfter anyway when checking if the pointer aliases any of the Call's arguments:
https://github.com/llvm/llvm-project/blob/6c143a86cddbc6d0431dd643bfc7d4f017042512/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L973-L978